src/runtime/Codecs/FunctionCodec.cs

```diff
@@ -0,0 +1,377 @@
+            }
+        }
+    }
+
+    internal class CallableHelper
+    {
+        public static object[] ConvertArgs(IntPtr args)
+        {
+            var numArgs = Runtime.PyTuple_Size(args);
+            object[] managedArgs = null;
+            if (numArgs > 0)
+            {
+                managedArgs = new object[numArgs];
+                for (int idx = 0; idx < numArgs; ++idx)
+                {
+                    IntPtr item = Runtime.PyTuple_GetItem(args, idx);
+                    object result;
+                    if (!Converter.ToManaged(item, typeof(object), out result, false))
+                        throw new Exception();
+
+                    managedArgs[idx] = result;
+                }
+            }
+            return managedArgs;
+        }
+    }
+
+    interface ICallable
+    {
+        IntPtr RunAction(IntPtr self, IntPtr args);
+    }
+
+    //class which wraps a thing
+    internal class ActionWrapper : ICallable
+    {
+        Action<object[]> _action = null;
+        internal ActionWrapper(Action action)
+        {
+            _action = (object[] args) => { action(); };
+        }
+
+        internal ActionWrapper(Action<object[]> action)
+        {
+            _action = action;
+        }
+
+        public virtual IntPtr RunAction(IntPtr self, IntPtr args)
+        {
+            try
+            {
+                object[] managedArgs = CallableHelper.ConvertArgs(args);
+                _action(managedArgs);
+            }
+            catch
+            {
+                Exceptions.SetError(Exceptions.TypeError, "Error in RunAction");
+                return IntPtr.Zero;
+            }
+
+            return Runtime.PyNone;
+        }
+    }
+
+    internal class FunctionWrapper : ICallable
+    {
+        Func<object[], object> _func = null;
+
+        internal FunctionWrapper(Func<object> func)
+        {
+            _func = (object[] args) => { return func(); };
+        }
+
+        internal FunctionWrapper(Func<object[], object> func)
+        {
+            _func = func;
+        }
+
+        public virtual IntPtr RunAction(IntPtr self, IntPtr args)
+        {
+            try
+            {
+                object[] managedArgs = CallableHelper.ConvertArgs(args);
+                var retVal = _func(managedArgs);
+                return Converter.ToPython(retVal);
+            }
+            catch
+            {
+                Exceptions.SetError(Exceptions.TypeError, "Error in RunAction");
+                return IntPtr.Zero;
+            }
+        }
+    }
+
+    //converts python functions to C# actions
+    class FunctionCodec : IPyObjectDecoder, IPyObjectEncoder
+    {
+        private static int GetNumArgs(PyObject pyCallable)
+        {
+            var locals = new PyDict();
+            locals.SetItem("f", pyCallable);
+            using (Py.GIL())
+                PythonEngine.Exec(@"
+from inspect import signature
+x = len(signature(f).parameters)
```

Comment:
AFAIK, `signature` might fail.

Reply:
@lostmsu i tried to find something in the C API but failed.  I can try putting this in a try/except and return 0 in the except block